lib/rspec/rails/matchers/send_email.rb

```diff
@@ -0,0 +1,123 @@
+module RSpec
+  module Rails
+    module Matchers
+      # @api private
+      #
+      # Matcher class for `send_email`. Should not be instantiated directly.
+      #
+      # @see RSpec::Rails::Matchers#send_email
+      class SendEmail < RSpec::Rails::Matchers::BaseMatcher
+        INSPECT_EMAIL_ATTRIBUTES = %i[subject from to cc bcc].freeze
+
+        def initialize(criteria)
+          @criteria = criteria
+        end
+
+        # @private
+        def supports_value_expectations?
+          false
+        end
+
+        # @private
+        def supports_block_expectations?
+          true
+        end
+
+        def matches?(block)
+          define_matched_emails(block)
+
+          @matched_emails.one?
+        end
+
+        private
+
+        def diffable?
+          true
+        end
+
+        def match_when_negated(block)
+          define_matched_emails(block)
+          @matched_emails.empty?
+        end
+
+        def failure_message
+          result =
+            if multiple_match?
+              'More than 1 matching emails were sent.'
+            else
+              'No matching emails were sent.'
+            end
+          "#{result}#{sent_emails_message}"
+        end
+
+        def failure_message_when_negated
+          "Expected not to send an email but it was sent."
```

Comment:
Freeze string literals. You are doing it already for array `INSPECT_EMAIL_ATTRIBUTES`. Makes sense to do it here too. Or declare modeline to freeze them automatically.

Reply:
I see `freeze` is used occasionally in other places, so removing it from the const array declaration `INSPECT_EMAIL_ATTRIBUTES`. Freezing it or not is not very important, as it's highly unlikely that someone would mutate it. Freezing strings is not common in other similar places within this project. Therefore, this question is related to the project as a whole rather than just this pull request. The project uses Rubocop, and if configured properly, it would help eliminate this question as well as the one mentioned above.

Reply:
Adding the magic freeze string literals to the top of the file is certainly fine. We have a very custom rubocop check list in part due to legacy history, a lot of our code predates Rubocop's existence. (Rubocop's RSpec rules are also not endorsed by us in anyway).

Reply:
Ok, changing to freezing literals. Rubocop can be configured to validate new files (the rest files can be added to "rubocop todo"), but it's a separate story.